### PR TITLE
Add LocationChangedAction handler and tests to support the Windows Compat module.

### PR DESCRIPTION
**Pull request by BrucePay**
## PR Summary

Fix for #7551. Implemented a new action `LocationChangedAction ` that fires when the current directory in a runspace is changed.

## PR Checklist

- [x] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [x] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [x] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [x] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [ ] Not Applicable
    - **OR**
    - [x] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [ ] Not Applicable or can only be tested interactively
    - **OR**
    - [x] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

### CODE REVIEW

**Review by iSazonov — COMMENTED** [review]
@BrucePay Please address CodeFactor issues for your new code and public comments.

File: test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1 [hunk 1]
```diff
@@ -144,4 +144,40 @@ Describe "Set-Location" -Tags "CI" {
...
+            $eventSessionState = $null
+            $eventRunspace = $null
+            $ExecutionContext.InvokeCommand.LocationChangedAction = {
+                (Get-Variable eventRunspace).Value = $this
+                (Get-Variable eventSessionState).Value = $_.SessionState
+                (Get-Variable oldPath).Value = $_.oldPath
+                (Get-Variable newPath).Value = $_.newPath
+            }
+            Set-Location ..
+            $newPath.Path | Should Be $pwd.Path
```
[the remark this answers is not in the document]
Comment (BrucePay) [reply test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1]: Oops. Yeah - I'll fix that. Thanks.

**Review by rjmholt — APPROVED** [review]
This looks good to me, modulo @iSazonov's Pester comment

File: test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1 [hunk 2]
```diff
@@ -144,4 +144,40 @@ Describe "Set-Location" -Tags "CI" {
...
+            $eventSessionState | Should -Be $ExecutionContext.SessionState
+            $eventRunspace | Should -Be ([runspace]::DefaultRunspace)
+        }
+
+        It 'Errors in the LocationChangedAction should be catchable but not fail the cd' {
+            $location = $PWD
+            Set-Location ..
+            $ExecutionContext.InvokeCommand.LocationChangedAction = { throw "Boom" }
+            # Verify that the exception occurred
+            { Set-Location $location } | Should Throw "Boom"
```
Comment (rjmholt) [inline test/powershell/Modules/Microsoft.PowerShell.Management/Set-Location.Tests.ps1]: Ideally this should be `Should -Throw -ErrorId "Boom"` here

**Review by iSazonov — APPROVED** [review]
LGTM with one comment about `-Throw`.

**Comment by anmenaga** [comment]
@TravisEz13 This looks read for merge.
Please note that it has "User-facing Documentation needed".

**Pull request merged**